In [1]:
import pandas as pd

DATA_PATH = '../data/'
DATA_INTERIM_PATH = DATA_PATH + 'interim/'

train = pd.read_csv(DATA_INTERIM_PATH + 'train.csv')
val = pd.read_csv(DATA_INTERIM_PATH + 'val.csv')

In [18]:
train.head()

id published-at                                              title  \
0   1   2017-09-10  After DeVos Announced Plans To Reexamine Title...   
1   2   2017-05-05  University To Award Trayvon Martin With Posthu...   
2   8   2017-11-15  Texas State University suspends Greek life aft...   
3   9   2015-06-10  Red Sox waste Rodriguez outing in 1-0 loss to ...   
4  10          NaN                          Eve and the New Jerusalem   

                                        article_text  \
0  After DeVos Announced Plans To Reexamine Title...   
1  University To Award Trayvon Martin With Posthu...   
2  Texas State University suspends Greek life aft...   
3  Red Sox waste Rodriguez outing in 1-0 loss to ...   
4  Eve and the New Jerusalem Imaginings of a utop...   

                                      external_links internal_links  \
0  ['https://www.nbcnews.com/news/us-news/betsy-d...             []   
1       ['https://www.facebook.com/FLMemorialUniv/']             []   
2  ['https://www.nbcnews.com/storyline/hazing-in-...             []   
3                                                 []             []   
4                                                 []             []   

   hyperpartisan          bias  \
0           True         right   
1           True         right   
2          False  right-center   
3          False         least   
4           True          left   

                                                 url labeled-by  
0  https://dailywire.com/news/20858/after-devos-a...  publisher  
1  https://dailywire.com/news/16103/university-aw...  publisher  
2  https://newsline.com/texas-state-university-su...  publisher  
3  https://apnews.com/795c45f6db904f8eb85d72cee61...  publisher  
4  https://dissentmagazine.org/article/eve-and-th...  publisher

Average external links

In [3]:
import ast
len(ast.literal_eval(train['external_links'][0]))

7

In [20]:
e_links_count_HP = 0
for links in train[train['hyperpartisan']==True]['external_links']:
    e_links_count_HP += len(ast.literal_eval(links))
    
e_links_count_HP/train[train['hyperpartisan']==True].shape[0]

3.3713375

In [12]:
e_links_count_nonHP = 0
for links in train[train['hyperpartisan']==False]['external_links']:
    e_links_count_nonHP += len(ast.literal_eval(links))
    
e_links_count_nonHP/train[train['hyperpartisan']==False].shape[0]

2.0891625

In [13]:
i_links_count_HP = 0
for links in train[train['hyperpartisan']==True]['internal_links']:
    i_links_count_HP += len(ast.literal_eval(links))
    
i_links_count_HP/train[train['hyperpartisan']==True].shape[0]

0.1684775

In [15]:
i_links_count_nonHP = 0
for links in train[train['hyperpartisan']==False]['internal_links']:
    i_links_count_nonHP += len(ast.literal_eval(links))
    
i_links_count_nonHP/train[train['hyperpartisan']==False].shape[0]

0.117455

HP vs. nonHP

TRAIN:

3.37 vs. 2.09 external links on average

0.17 vs. 0.12 internal links on average

VAL:

4.80 vs. 6.55 external links on average

In [19]:
e_links_count_HP = 0
for links in val[val['hyperpartisan']==True]['external_links']:
    e_links_count_HP += len(ast.literal_eval(links))
    
e_links_count_HP/val[val['hyperpartisan']==True].shape[0]

4.80703

In [21]:
e_links_count_nonHP = 0
for links in val[val['hyperpartisan']==False]['external_links']:
    e_links_count_nonHP += len(ast.literal_eval(links))
    
e_links_count_nonHP/val[val['hyperpartisan']==False].shape[0]

6.55377

In [23]:
i_links_count_HP = 0
for links in val[val['hyperpartisan']==True]['internal_links']:
    i_links_count_HP += len(ast.literal_eval(links))
    
i_links_count_HP/val[val['hyperpartisan']==True].shape[0]

0.02928

In [24]:
i_links_count_nonHP = 0
for links in val[val['hyperpartisan']==False]['internal_links']:
    i_links_count_nonHP += len(ast.literal_eval(links))
    
i_links_count_nonHP/val[val['hyperpartisan']==False].shape[0]

5.96074

In [26]:
train['num_external_links'] = train.loc[:,'external_links'].apply(lambda x: len(ast.literal_eval(x)))

In [28]:
train[['external_links', 'num_external_links']].head()

external_links  num_external_links
0  ['https://www.nbcnews.com/news/us-news/betsy-d...                   7
1       ['https://www.facebook.com/FLMemorialUniv/']                   1
2  ['https://www.nbcnews.com/storyline/hazing-in-...                   2
3                                                 []                   0
4                                                 []                   0

In [29]:
DATA_PROCESSED_PATH = DATA_PATH + 'processed/'
train = pd.read_csv(DATA_PROCESSED_PATH + 'train.csv')
val = pd.read_csv(DATA_PROCESSED_PATH + 'val.csv')

### Extracting domains

In [30]:
import tldextract

In [33]:
tldextract.extract('https://www.nbcnews.com/news/')

ExtractResult(subdomain='www', domain='nbcnews', suffix='com')

In [36]:
train['domain'] = train.loc[:,'url'].apply(lambda x: tldextract.extract(x)[1])

In [37]:
train.head()

id published-at                                              title  \
0   1   2017-09-10  After DeVos Announced Plans To Reexamine Title...   
1   2   2017-05-05  University To Award Trayvon Martin With Posthu...   
2   8   2017-11-15  Texas State University suspends Greek life aft...   
3   9   2015-06-10  Red Sox waste Rodriguez outing in 1-0 loss to ...   
4  10          NaN                          Eve and the New Jerusalem   

                                        article_text  \
0  After DeVos Announced Plans To Reexamine Title...   
1  University To Award Trayvon Martin With Posthu...   
2  Texas State University suspends Greek life aft...   
3  Red Sox waste Rodriguez outing in 1-0 loss to ...   
4  Eve and the New Jerusalem Imaginings of a utop...   

                                      external_links internal_links  \
0  ['https://www.nbcnews.com/news/us-news/betsy-d...             []   
1       ['https://www.facebook.com/FLMemorialUniv/']             []   
2  ['https://www.nbcnews.com/storyline/hazing-in-...             []   
3                                                 []             []   
4                                                 []             []   

   hyperpartisan          bias  \
0           True         right   
1           True         right   
2          False  right-center   
3          False         least   
4           True          left   

                                                 url labeled-by  \
0  https://dailywire.com/news/20858/after-devos-a...  publisher   
1  https://dailywire.com/news/16103/university-aw...  publisher   
2  https://newsline.com/texas-state-university-su...  publisher   
3  https://apnews.com/795c45f6db904f8eb85d72cee61...  publisher   
4  https://dissentmagazine.org/article/eve-and-th...  publisher   

   num_external_links  num_internal_links           domain  
0                   7                   0        dailywire  
1                   1                   0        dailywire  
2                   2                   0         newsline  
3                   0                   0           apnews  
4                   0                   0  dissentmagazine

In [38]:
train['domain'].value_counts()

foxbusiness                  116106
abqjournal                    93853
apnews                        75725
pri                           51781
newsline                      47593
counterpunch                  42410
motherjones                   38487
truthdig                      33780
dailywire                     21538
thedailybeast                 19560
poynter                       17276
baptistnews                   13728
nbcnews                       12829
reuters                       10914
consortiumnews                 9593
washingtonblade                8265
calwatchdog                    7611
thegoldwater                   6947
natmonitor                     6556
dissentmagazine                6292
politicalillusionsexposed      5714
chicagoreporter                5316
ivn                            4967
eppc                           4652
mintpressnews                  4494
factcheck                      4485
billmoyers                     4328
circa                       